static/js/recent_senders.js

```diff
@@ -1,91 +1,147 @@
+import _ from "lodash";
+
 import {FoldDict} from "./fold_dict";
-import * as message_util from "./message_util";
+import * as message_store from "./message_store";
 
-// topic_senders[stream_id][topic_id][sender_id] = latest_message_id
-const topic_senders = new Map();
-// topic_senders[stream_id][sender_id] = latest_message_id
+class IdList {
+    ids = [];
+    _max_id = undefined;
+
+    add(id) {
+        this.ids.push(id);
+        this._max_id = _.max(this.ids);
+    }
```

Comment:
I'm worried that when we fetch from the server 10K messages sent by a bot to the same topic, it'll make the browser hang, with this new algorithm.

Reply:
It won’t.

Reply:
But we can make max_id lazy.

Reply:
The amount of work we do (including browser work) to process a message dwarfs finding the max of 10k integers.

Reply:
Yeah, lazy fixes the case I had; now you can only get to quadratic work if you have some workload that alternates doing compose typeahead things that need to read this with adding new messages, which seems pretty unlikely.

That said ... I guess a Data Structures textbook would tell us what we actually want is for `IdList` to be replaced with a max-heap?  

I don't want to add a heap implementation to the codebase for this, but surely there's a tested JS library we can just use.

@andersk FYI

Reply:
If you have a single runaway bot, the first keystroke you type finds the max of 10k integers in a set.  Then everything is cached.  If the bot continues sending messages, subsequent keystrokes still do O(1) work for that bot, since, as I mention in the commit message, we don't invalidate the cached max value for the add case.  If somebody was rapidly deleting messages while you were typing, then you'd hit O(N) work every time, but it still wouldn't be quadratic.

Reply:
It's also worth mentioning again how bad the current situation is on master.  On master if you are narrowed to a **different** stream than the runaway bot, you pay O(N) costs for a bigger N.  Literally every time we update/delete a message on master, we sweep the entire list of messages to rebuild the data structure.

Reply:
(Ah, one thing that I didn't realize is that Tim's comments preceded my latest push by about an hour. See my latest fix to see how I address incremental adds.)